Add WebXR support (for Godot 3.2)

dsnopek opened the pull request:
Implements proposal https://github.com/godotengine/godot-proposals/issues/1581

This is for Godot 3.2, because WebXR depends or WebGL, and there is presently no OpenGL/WebGL support in the master branch. When it's added back, I can port this to master.

**This is a work-in-progress!** Once it's fully functional, it will get lots of clean-up and better error handling, and I'll squash it or start a clean pull request.

I'm hoping that I can use this PR to get some review and collaborate with other contributors. :-)

dsnopek committed:
Add support for WebXR

dsnopek commented:
Here's a YouTube video demonstrating what's working so far:

https://www.youtube.com/watch?v=UKHjR60i5jI

Unfortunately, at the moment, I'm stuck on figuring out how to render the render targets to the headset (rather than the screen). I explain the problem towards the end of the video, but I'll attempt to explain it here as well:

WebXR passes us a framebuffer object and viewport dimensions that it wants us to render each eye to, in order to send the data to the headset. Ideally, I'd like to bind to that framebuffer, and then blit the render target to the framebuffer in the given viewport for each eye.

Calling `VSG::rasterizer->blit_render_target_to_screen(p_render_target, viewport, 0);` almost does what I want, but it binds to the screen's framebuffer before blitting, so that doesn't get the data to the headset. In my PR, I copied that method, making a new `VSG::rasterizer->blit_render_target_to_current_framebuffer(p_render_target, viewport);` where I removed the part where it calls `glBindFramebuffer()` but that doesn't work for some reason.

If anyone has any advice on how to implement that method, or any other better ways to do this, please let me know!

@BastiaanOlij you're the AR/VR expert - any thoughts?

Thanks in advance! :-)

HeadClot commented:
Good work so far @dsnopek. Got a question though :)

Will we be able to use this along side plugins such as [Qodot](https://github.com/Shfty/qodot-plugin) for example?

dsnopek commented:
@HeadClot I don't see why not! If you can use Qodot in games exported to HTML5, you should still be able to use it with WebXR enabled :-)

One important note, though: this isn't a plugin, but a change to Godot core. So, until this gets merged into Godot and released (which I hope it does, but that may never happen), in order to use this, you'd have to patch and compile the HTML5 export template yourself (or use the template that someone else already compiled with this patch)

HeadClot commented:
> @HeadClot I don't see why not! If you can use Qodot in games exported to HTML5, you should still be able to use it with WebXR enabled :-)
> 
> One important note, though: this isn't a plugin, but a change to Godot core. So, until this gets merged into Godot and released (which I hope it does, but that may never happen), in order to use this, you'd have to patch and compile the HTML5 export template yourself (or use the template that someone else already compiled with this patch)

Thank you for the information about Qodot. And totally Understood about the Pull Request. :)

fire commented:
Are you able to figure out why the Javascript platform is failing?

dsnopek commented:
> Are you able to figure out why the Javascript platform is failing?

I'm not sure yet. Here's the actual error from the CI log:

```
[JSC_UNDEFINED_VARIABLE] variable XRWebGLLayer is undeclared
```

XRWebGLLayer is class that should be defined by the browser.

Perhaps its a difference between the Emscripten version that I'm using locally versus the version used by CI?

I will certainly make sure all the tests are passing before taking the PR out of "draft" status. :-)

Faless commented:
@dsnopek probably need to be added as an extern for the closure compiler. I'd say don't worry too much about it, I can help you fix it when you feel the PR is ready.

dsnopek commented:
@Faless Ok, great, thanks!

m4gr3d commented:
@dsnopek Great work! Leaving it out to @BastiaanOlij to confirm, but it seems that with the exposure of a couple functions to `gdnative`, this can be turn into an external gdnative plugin similar to the ones in https://github.com/GodotVR.

Do you think there is a strong reason to have it within the engine itself as a module?

dsnopek commented:
@m4gr3d Thanks!

> [...] it seems that with the exposure of a couple functions to gdnative, this can be turn into an external gdnative plugin similar to the ones in https://github.com/GodotVR. Do you think there is a strong reason to have it within the engine itself as a module?

I mentioned this in the proposal (https://github.com/godotengine/godot-proposals/issues/1581): WebXR is for HTML5 only, and, unfortunately, gdnative doesn't work in the HTML5. This may also require some minor changes to the HTML5 export, so at least some part of this needs to be done in Godot core.

m4gr3d commented:
> @m4gr3d Thanks!
> 
> > [...] it seems that with the exposure of a couple functions to gdnative, this can be turn into an external gdnative plugin similar to the ones in https://github.com/GodotVR. Do you think there is a strong reason to have it within the engine itself as a module?
> 
> I mentioned this in the proposal ([godotengine/godot-proposals#1581](https://github.com/godotengine/godot-proposals/issues/1581)): WebXR is for HTML5 only, and, unfortunately, gdnative doesn't work in the HTML5. This may also require some minor changes to the HTML5 export, so at least some part of this needs to be done in Godot core.

@dsnopek Sounds good, thanks for the clarification!

dsnopek commented:
My latest commit gets this actually displaying to the headset on an Oculus Quest using GLES2! However, something is wrong with the eye transform or projection, because it doesn't give the correct binocular effect. And GLES3 doesn't work for some reason either.

But it's pretty exciting to finally get it displaying something on a real headset! Once it's looking a little better I'll record a video from the headset :-)

BastiaanOlij commented:
I have to dive into WebXR a little more but like I said in the comment on the Youtube video, there has to be a way for WebXR to simply provide the texture buffer they want you to render into. 
That would do away with the extra copy you are doing right now and possibly solve some of the other issues you are having.

BastiaanOlij commented on modules/webxr/config.py:
```diff
@@ -0,0 +1,15 @@
+def can_build(env, platform):
+    return True
```
Make sure you check your platform here so the module only gets compiled for HTML5

dsnopek replied:
Fixed in latest commit!

dsnopek replied:
Ended up going back to the original here, so that we could compile in the abstract class and get autocomplete in the editor.

BastiaanOlij commented on modules/webxr/webxr_interface.cpp:
```diff
@@ -0,0 +1,328 @@
+		return transform_for_eye;
+	}
+
+	int view_index = (p_eye == ARVRInterface::EYE_RIGHT) ? 1 : 0;
```
Note, if p_eye is EYE_MONO you need to return the center position between the two eyes. this is where the camera object is placed within the node tree. While not used for rendering there are other parts that use the nodes transform to position other object or query this info to get the players head position.

dsnopek replied:
Fixed in the latest commit!

dsnopek commented:
@BastiaanOlij Thanks for the code review! I'll try to address both of those things later today.

> I have to dive into WebXR a little more but like I said in the comment on the Youtube video, there has to be a way for WebXR to simply provide the texture buffer they want you to render into.

I spent quite a bit of time yesterday trying to see if there was a way to get a _texture_, I couldn't find anything.

The `XRWebGLLayer'` object is what connects connects WebXR to WebGL:

https://developer.mozilla.org/en-US/docs/Web/API/XRWebGLLayer

It only gives us access to a `WebGLFramebuffer`, not a texture, and I couldn't find anything in the WebGL API that would allow us to get the texture object from the framebuffer. That said, I'm really not that great at OpenGL/WebGL, so perhaps there is some way that a more experienced OpenGL developer would know?

However, if we only have a framebuffer, I wonder if we could instead add a new method to Godot's `RasterizerStorage` like `render_target_set_external_framebuffer()` to allow having a render target that has a framebuffer without an associated texture?

Then we'd also need to modify the `ARVRInterface` class to have an new optional virtual method like `get_external_framebuffer_for_eye()' that would allow the `WebXRInterface` class to return a framebuffer that could be assigned to the render target and allow Godot to render directly to it?

Seems like some big changes to Godot's internals, though, just for WebXR...

dsnopek commented:
The latest changes get a lot closer to correctly rendering!

I ended up realizing that the main problem was that the code for blitting the textures to the headset's framebuffer was stretching the texture in a weird way. So, I ended up doing the thing I really didn't want to do: writing some custom WebGL code to render the texture on a quad with a little custom shader and everything. But that solved it, got it working on both GLES2 and GLES3, and allowed me to remove the changes I made to the rasterizer.

However, it's still not quite right... I think the `get_transform_for_eye()` is off somehow. It feels like it's rotating around the wrong point or something? Like, rotating my head seems to move me in space in an unnatural way. Testing while it has these problems has been making me feel quite sick, so probably time to stop for the night. :-)

BastiaanOlij commented:
Try flipping rows and columns on the matrix, so something like:

For (j=0;j<4;j++) {
For (i=0;i<4;i++) {
Mnew[j][i] = Mold[i][j];
}
}

(Sorry on my phone so but hopefully you get the idea)

On Sat, 3 Oct 2020 at 12:13 pm, David Snopek <notifications@github.com>
wrote:

>
>
> The latest changes get a lot closer to correctly rendering!
>
>
> I ended up realizing that the main problem was that the code for blitting
> the textures to the headset's framebuffer was stretching the texture in a
> weird way. So, I ended up doing the thing I really didn't want to do:
> writing some custom WebGL code to render the texture on a quad with a
> little custom shader and everything. But that solved it, got it working on
> both GLES2 and GLES3, and allowed me to remove the changes I made to the
> rasterizer.
>
>
> However, it's still not quite right... I think the get_transform_for_eye()
> is off somehow. It feels like it's rotating around the wrong point or
> something? Like, rotating my head seems to move me in space in an unnatural
> way. Testing while it has these problems has been making me feel quite
> sick, so probably time to stop for the night. :-)
>
>
>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/42397#issuecomment-703030715>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO262PYJQMY4PYQP2GPCGLSI2CDPANCNFSM4R5G635A>
> .
>
>
> --
Kindest regards,

Bastiaan Olij

My personal profile on Facebook <https://www.facebook.com/bastiaan.olij>
My professional profile on LinkedIn
<https://www.linkedin.com/in/bastiaanolij/>
My profile on Twitter <https://twitter.com/mux213>
My tutorials on Youtube <https://www.youtube.com/BastiaanOlij>
My code repositories on Github <https://github.com/BastiaanOlij>

dsnopek commented:
@BastiaanOlij You are right! Of course, I didn't see your comment until just now, after I'd already found the same solution. :-)

I actually tried swapping column-major vs row-major more than once earlier on, but I think some other bug(s) that I've since solved, must have prevented that from working (or at least showing a noticeable difference) at the time.

So, with the commit I just pushed, it's actually rendering fully correctly as far as I can tell on my Oculus Quest!

There's still lots to do, of course: some clean-up, some API additions so the developer can initialize WebXR with the desired parameters, etc. And I'd really like to at least try experimenting with AR (since VR is only half of WebXR), but I have no previous experience with AR, so that might take me a bit to figure out.

dsnopek commented:
I just recorded a new video which shows this PR running on a real Oculus Quest:

https://www.youtube.com/watch?v=a5IHo_MLC5c

HeadClot commented:
> I just recorded a new video which shows this PR running on a real Oculus Quest:
> 
> https://www.youtube.com/watch?v=a5IHo_MLC5c

Oh wow! Did not expect this to come this far so soon. Good work!
Cannot wait for controllers to be working :)

If you do not mind me asking @dsnopek what parameters will you be exposing to devs?

dsnopek commented:
@HeadClot:

> If you do not mind me asking @dsnopek what parameters will you be exposing to devs?

Namely, allowing devs to call `XRSystem.isSessionSupported()` themselves, and then to choose the session mode and reference space type, based on what the browser says is supported. There may be a couple other little things, but those are the big ones.

dsnopek commented:
In my latest commit, I added some APIs for configuring how WebXR should be initialized, and I'd love some input from Godot core developers on how those APIs look.

I have a couple of concerns:

- I used strings (because that's what they are in WebXR/Javascript) but using enums would feel more _Godot-ish_, especially because they could theoretically be autocompleted in the editor. However, I worry that using enums could be too limiting if WebXR changes (say, adds some new allowed values), so sticking with strings feels a little more future-proof. But, also, since this module is only compiled on the 'javascript' platform, the editor wouldn't actually know about any enums that I added, because the types wouldn't be registered. Is there a way around that?
- For parameters that take an array of strings, I used a single string with commas separating the values. I originally made the properties as `Vector<String>`, but I just couldn't get an array of strings to pass from C++ to Javascript. However, many of these parameters are also the kind of thing where you'd usually put in one value, but maybe, optionally, have multiple, so this sort of makes sense in its own right. But I have mixed feelings about it, and wondered if anyone else had any thoughts?
- Related to the first point, this adds parameters, methods and signals to `WebXRInterface` and in my demo project I'm able to use them perfectly fine. However, since this module is only compiled on the 'javascript' platform, the editor doesn't know about any of these, and so can't do autocomplete. Is there a way around that? Should I even be worried about it?

Thanks in advance for your thoughts! :-)

dsnopek commented:
@Faless Thanks so much for the advice! I implemented most of it (see my first 2 points below). :-)

In my latest bunch of commits:

- **I moved away from `EM_ASM` into an Emscripten library.** The code is much easier to understand, it got rid of all the ugliness with `Module` and now it's getting access to the `Browser` and `GL` objects the normal way, instead of the weird way I was doing previously.
- **I separated the `WebXRInterface` into an abstract class that's always compiled, and an implementation (`WebXRInterfaceJS`) for the 'javascript' platform.** So, now WebXRInterface and its properties, methods and signals are autocompleted in the editor.
- **There's positional tracking of controllers.** Unfortunately, inputs (buttons, joystick) on the controllers don't yet work. Since `WebXR` gives a gamepad object, I was hoping that all we'd need to do is pass the index for it to `ARVRPositionalTracker::set_joy_id()` and let `ARVRController` take care of all that. However, the [WebXR spec dictates](https://immersive-web.github.io/webxr-gamepads-module/#navigator-differences) that XR controllers must NOT be registered as gamepads the normal way, and hence aren't returned by `navigator.getGamepads()` and always have an index of -1. I'm not sure how to handle this yet - it's what I'll be spending next week trying to figure out!

Regarding using enums in the API instead of strings: I got most of the way through implementing that, but ended up deciding that I think the strings are better. In GDScript, this is how using enums would have looked:

```
webxr_interface.session_mode = WebXRInterface.SESSION_IMMERSIVE_VR
webxr_interface.required_features = [WebXRInterface.REFERENCE_SPACE_LOCAL_FLOOR]
webxr_interface.optional_features = [WebXRInterface.REFERENCE_SPACE_BOUNDED_FLOOR]
webxr_interface.requested_reference_space_types = [
	WebXRInterface.REFERENCE_SPACE_BOUNDED_FLOOR,
	WebXRInterface.REFERENCE_SPACE_LOCAL_FLOOR,
	WebXRInterface.REFERENCE_SPACE_LOCAL,
]
```

Having to use arrays for `required_features` and `optional_features` feels weird and easy to forget. Also, those values currently only take reference spaces, but [the spec explicitly states](https://www.w3.org/TR/webxr/#dom-xrsessioninit-requiredfeatures) that other values could be added in the future. It also just seems like it'd be easier to document this by directly exposing the same strings that are in the WebXR spec and that would be discussed elsewhere online.

But I still have the unfinished enum implementation in a branch, in case I change my mind later. :-)

dsnopek commented:
Here's my next video progress report, mostly discussing why getting the controller inputs (buttons & joystick) to work is challenging (but it also shows the positional tracking of controllers working on my Quest!):

https://www.youtube.com/watch?v=2HKfpWT5opU

dsnopek commented:
@BastiaanOlij I've been trying to figure out how to handle the situation with controller inputn (as described in my video and the last two comments). I'm starting to think that the best way is going to be registering the XR gamepads with the input singleton using arbitrarily high ids, like 100 and 101 (for the left and right controller), so that they don't conflict with normal gamepads (whose ids start at 0). Does this sound like an OK plan? Or is this a terrible idea?

There’s a couple things that worry me about this plan:

- Device ids on the “Input map” in project settings only go up to device id 7. This mostly matters for the joystick axes, because you can get the buttons from the signals on the `ARVRController` node. But you could also get the joystick events from the _input() method, rather than using actions on the “Input map”, so maybe that doesn’t matter? 
- Making a Godot project that’s compatible with WebXR and, for example, Oculus Mobile (Quest) will be tricker because the gamepads on Oculus Mobile will have device ids 0 and 1. (Although, if you’re not using the “Input map” and look at the joystick events directly, checking the platform and adding 100 to the ids probably isn’t that big of a deal, so maybe this doesn’t matter either?)

However, looking at how the `ARVRController` class is emitting the "button_pressed" and "button_released" signals, I don't see any other way for that to work besides having a valid joy id on the `ARVRPositionalTracker`. And I can’t think of a way for my `WebXRInterface` to emit those signals itself, because it doesn’t know anything about those nodes.

But if registering the gamepads with fake ids isn’t a good solution, then the only other idea I have is putting some special signals on `WebXRInterface` which developers would need to use to get buttons and joysticks just for this platform. Or are there some other idea(s) I’m not thinking of?

Thanks in advance for your thoughts!

dsnopek commented:
This _might_ be a terrible idea, but... :-) my latest commit implements the plan I described above where `WebXRInterface` will register some "virtual" gamepads at ids 100 and 101 for the left and right controller. This works in my testing for `ARVRController` to emit it's 'button_pressed' and 'button_released' signals, and getting the joystick axes as `InputEventJoystickMotion` events from `_input()`.

dsnopek commented:
Here's my latest progress report video, showing a new demo which gets controller inputs:

https://www.youtube.com/watch?v=uy6moGOIBmQ

ajlennon commented:
Great work @dsnopek. We're really interested in the intersection of Godot and WebXR for the kinds of reasons you've talked about with the "walled garden" and enabling VR users to get going without complex install procedures.

I'll have a go at compiling with your PR and see what I can get going on our Oculus Quest 2 and Rift S here if that's of interest

dsnopek commented:
I just wrote a tutorial about how to use this, including posting some pre-compiled builds of the HTML5 export templates that work with Godot 3.2.3, so you can test it without having to compile anything:

https://www.snopekgames.com/blog/2020/how-make-vr-game-webxr-godot

ajlennon commented:
Wonderful thanks @dsnopek - will work through here :)

dsnopek commented:
@Faless I've been trying to fix the closure compiler error that's causing the CI build to fail, and I haven't managed to, so I'd like to take you up on your earlier offer to help. :-)

I'm trying to follow these docs from the Closure Compiler:

https://developers.google.com/closure/compiler/docs/externs-and-exports

In my latest commit, you can see the code that I added to `platform/javascript/engine/externs.js`:

https://github.com/godotengine/godot/pull/42397/commits/9efbb928347563d06b418fbff81de2332e3b1dc5#diff-3394e72f4930953adc5cedca9d23f9feb912f5d71ac10e2939c0ad1e9a3d5d45R5

Unfortunately, even with that code added, I'm still getting:

```
ERROR - [JSC_UNDEFINED_VARIABLE] variable XRWebGLLayer is undeclared
```

Any suggestions?

Faless commented:
Sure. I'll check out the branch and have a look.

Faless commented:
As discussed on IRC, sadly the webxr externs are missing, here is a partial fix (apply on top of your branch).
Add two functions to build script: `addJSLibraries` and `addJSExterns`.
Your module now uses those functions and have an extra file: `webxr_externs.js` (with example constructor, properties, return value, type definition).

```diff
diff --git a/modules/webxr/native/webxr_externs.js b/modules/webxr/native/webxr_externs.js
new file mode 100644
index 0000000000..bafd33ecd6
--- /dev/null
+++ b/modules/webxr/native/webxr_externs.js
@@ -0,0 +1,54 @@
+/*
+ * @constructor
+ */
+function XRSession() {
+
+}
+
+/*
+ * @constructor
+ */
+function XRViewerPose() {
+
+}
+
+/*
+ * @constructor
+ */
+function XRFrame() {}
+
+/*
+ * @type {XRSession}
+ */
+XRFrame.prototype.session;
+
+/*
+ * @return {XRViewerPose?}
+ */
+XRFrame.prototype.getViewerPose = function(referenceSpace) {}
+
+/**
+ * @typedef {function(number, XRFrame): undefined}
+ */
+var XRFrameRequestCallback;
+
+/*
+ * @constructor
+ */
+function XRViewport() {}
+
+/*
+ * @constructor
+ */
+function XRView() {}
+
+/*
+ * @constructor
+ */
+function XRWebGLLayer(session, ctx) {}
+
+/*
+ * @param {XRView} view
+ * @return {XRViewport}
+ */
+XRWebGLLayer.prototype.getViewport = function(view) {};
diff --git a/platform/javascript/SCsub b/platform/javascript/SCsub
index 9db46b4ae6..8359a24fa8 100644
--- a/platform/javascript/SCsub
+++ b/platform/javascript/SCsub
@@ -17,21 +17,21 @@ if env["threads_enabled"]:
 
 build = env.add_program(build_targets, javascript_files)
 
-js_libraries = [
-    "native/http_request.js",
-    "native/library_godot_audio.js",
-]
-for lib in js_libraries:
-    env.Append(LINKFLAGS=["--js-library", env.File(lib).path])
-env.Depends(build, js_libraries)
+env.AddJSLibraries(
+    [
+        "native/http_request.js",
+        "native/library_godot_audio.js",
+    ]
+)
 
-js_pre = [
-    "native/id_handler.js",
-    "native/utils.js",
-]
-for js in js_pre:
-    env.Append(LINKFLAGS=["--pre-js", env.File(js).path])
-env.Depends(build, js_pre)
+for lib in env["JS_LIBS"]:
+    env.Append(LINKFLAGS=["--js-library", lib])
+env.Depends(build, env["JS_LIBS"])
+
+if "JS_EXTERNS" in env:
+    for ext in env["JS_EXTERNS"]:
+        env["ENV"]["EMCC_CLOSURE_ARGS"] += ' --externs ' + ext.path
+    env.Depends(build, env["JS_EXTERNS"])
 
 engine = [
     "engine/preloader.js",
diff --git a/platform/javascript/detect.py b/platform/javascript/detect.py
index 0ba9462cde..c8288b1e1d 100644
--- a/platform/javascript/detect.py
+++ b/platform/javascript/detect.py
@@ -1,6 +1,6 @@
 import os
 
-from emscripten_helpers import run_closure_compiler, create_engine_file
+from emscripten_helpers import run_closure_compiler, create_engine_file, add_js_libraries, add_js_externs
 from SCons.Util import WhereIs
 
 
@@ -95,11 +95,15 @@ def configure(env):
         jscc = env.Builder(generator=run_closure_compiler, suffix=".cc.js", src_suffix=".js")
         env.Append(BUILDERS={"BuildJS": jscc})
 
+    # Add helper method for adding libraries.
+    env.AddMethod(add_js_libraries, "AddJSLibraries")
+    env.AddMethod(add_js_externs, "AddJSExterns")
+
     # Add method that joins/compiles our Engine files.
     env.AddMethod(create_engine_file, "CreateEngineFile")
 
     # Closure compiler extern and support for ecmascript specs (const, let, etc).
-    env["ENV"]["EMCC_CLOSURE_ARGS"] = "--language_in ECMASCRIPT6 --externs " + env.File("#platform/javascript/engine/externs.js").path
+    env["ENV"]["EMCC_CLOSURE_ARGS"] = "--language_in ECMASCRIPT6"
 
     env["CC"] = "emcc"
     env["CXX"] = "em++"
diff --git a/platform/javascript/emscripten_helpers.py b/platform/javascript/emscripten_helpers.py
index f6db10fbbd..e516594550 100644
--- a/platform/javascript/emscripten_helpers.py
+++ b/platform/javascript/emscripten_helpers.py
@@ -19,3 +19,15 @@ def create_engine_file(env, target, source, externs):
     if env["use_closure_compiler"]:
         return env.BuildJS(target, source, JSEXTERNS=externs)
     return env.Textfile(target, [env.File(s) for s in source])
+
+
+def add_js_libraries(env, libraries):
+    if "JS_LIBS" not in env:
+        env["JS_LIBS"] = []
+    env.Append(JS_LIBS=env.File(libraries))
+
+
+def add_js_externs(env, externs):
+    if "JS_EXTERNS" not in env:
+        env["JS_EXTERNS"] = []
+    env.Append(JS_EXTERNS=env.File(externs))
```

Faless commented:
Ops, my bad, I forgot this bit:
```diff
diff --git a/modules/webxr/SCsub b/modules/webxr/SCsub
index 70f2ce2e0f..09e0ae1ad9 100644
--- a/modules/webxr/SCsub
+++ b/modules/webxr/SCsub
@@ -4,17 +4,8 @@ Import("env")
 Import("env_modules")
 
 if env["platform"] == "javascript":
-    # Copied from platform/javascript/SCsub so we can add dependencies.
-    javascript_build_targets = ["#bin/godot${PROGSUFFIX}.js", "#bin/godot${PROGSUFFIX}.wasm"]
-    if env["threads_enabled"]:
-        javascript_build_targets.append("#bin/godot${PROGSUFFIX}.worker.js")
-
-    js_libraries = [
-        "native/library_godot_webxr.js",
-    ]
-    for lib in js_libraries:
-        env.Append(LINKFLAGS=["--js-library", env.File(lib).path])
-    env.Depends(javascript_build_targets, js_libraries)
+    env.AddJSLibraries(["native/library_godot_webxr.js"])
+    env.AddJSExterns(["native/webxr_externs.js"])
 
 env_webxr = env_modules.Clone()
 env_webxr.add_source_files(env.modules_sources, "*.cpp")
```

dsnopek commented:
Thanks @Faless!

Just a note to those following along: the latest code on this PR is broken right now (@NeoSpark314 reported issues on the latest commit), because I need to build on the changes that @Faless added and fill in the `webxr_externs.js`. If you're looking to test before I fix that, you can download builds of the HTML5 export templates from Dropbox ([debug](https://www.dropbox.com/s/q769fj6f88avtwa/godot.javascript.opt.debug.zip?dl=1), [release](https://www.dropbox.com/s/p5c1gmyq58fnq27/godot.javascript.opt.zip?dl=1)) for Godot 3.2.3 that are based the PR at https://github.com/godotengine/godot/pull/42397/commits/9829722d05ff02419e2b87f7a6fcccec52df7c59 which should work.

dsnopek commented:
I've made a video version of my tutorial for how to use the WebXR support from this PR (for those who prefer video):

https://www.youtube.com/watch?v=vWP7Ti2nhKE

dsnopek commented:
@Faless I finally got it building and working with the Closure compiler! The IDL we had for WebXR was incomplete, but I attempted to add everything to the externs that's in the IDL, and then all the extra stuff that is used in this PR. Please take a look when you have a chance and let me know if I did anything wrong! Thanks :-)

dsnopek commented:
Here's my latest progress report video, mostly showing examples of app other folks have made for WebXR in Godot:

https://www.youtube.com/watch?v=BVVg8wvHcY0

ajlennon commented:
I've been having all sorts of fun!

https://dynamicdevices.github.io/ChristmasTreeLights

Faless commented:
@dsnopek wow, this is great :1st_place_medal:.

It still needs few style/binding fixes.
I will send you the diff to apply to speed up the process (I'm working on it).
The branch will also need a rebase which might conflict (I see you have a merge there): https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase (make sure to backup your files or create a backup branch to avoid losing any work).
I also think that another PR should be made on `master` (which only requires few renames to compile).
I know we can't test that, but at least the code will be there, with minimum changes, and it will be easier to hopefully keep the code update (plus we can follow the usual workflow).

I'll keep you posted on the diff.

ajlennon commented:
Not sure if this is the best place to ask but I would really like to get some video walls into a Godot project exported to WebXR.

I'd like these to be Youtube streams / livecams I serve out type of thing.

I have a video wall running in the above project from a local video but it appears Godot doesn't support URLs

https://github.com/godotengine/godot-proposals/issues/924

We were wondering if there might be a way to inject something into the exported WebXR to support this?

Any advice would be appreciated!

Faless commented:
Attached the first round of patches. You will need to do the rebase before if compiles (**make sure to backup you branch first**).

[patches.zip](https://github.com/godotengine/godot/files/5610716/patches.zip)

You can apply the patches via:
`git am *.patch`

If you are having trouble rebasing, ping me on IRC, I'll try to help you out.
You will have to apply another patch after the rebase to make the linter happy, but it won't apply before the rebase, so I'll sent it to you once that's done.

dsnopek commented:
@ajlennon In plain WebXR (meaning without using Godot), it's relatively easy to play a video from a URL. You can create a `<video>` HTML node pointing at a remote video, and use that to populate a WebGL texture. However, within Godot, I'm really not sure how you'd take advantage of that... I think there'd need to be some kind of special support in Godot's HTML5 export that allowed you to do that, but I really don't know exactly how that would need to work.

If this is a hard requirement for your project, it'd probably be a lot easier to implement using a game engine built specifically for WebXR, like [A-Frame](https://aframe.io/).

ajlennon commented:
> If this is a hard requirement for your project, it'd probably be a lot easier to implement using a game engine built specifically for WebXR, like [A-Frame](https://aframe.io/).

Very interesting - thanks for that @dsnopek

dsnopek commented:
@Faless Thanks so much!

> The branch will also need a rebase which might conflict (I see you have a merge there)

I was planning on rebasing this PR down to a single commit when it was actually ready to merge, which I was imagining would be after I have support for everything VR-related at the earliest. Until then, the Git history is actually pretty nice for me to have!

Or, do you think it's worth trying to get this merged in its current state and do additional PRs to finish it up? I've still got a few little things to add in order to complete support for everything VR-related. And then, of course, there's AR.

There hasn't really been a discussion here (or on the [proposal](https://github.com/godotengine/godot-proposals/issues/1581)) as to whether you and the other maintainers were OK with merging this into Godot core, so I've been kind of worried about that. :-)

Anyway, if you think this is ready to come out of "Draft" status and actually be considered for merging, I'll do the rebase and squash this down; otherwise, I think I'll hold off until that point, and just merge from 3.2 (which I've already done, so you can send the next patch you were saving).

> I also think that another PR should be made on master (which only requires few renames to compile).

Ok, once this is ready to come out of "Draft" status, I'll make another PR against master too.

Faless commented:
> Or, do you think it's worth trying to get this merged in its current state and do additional PRs to finish it up?

I guess that's also up to you. Do you feel it's already usable, and just missing some features? Or do you think it still needs some work to be usable?
Rebasing is not urgent, it was more a suggestion for you, since I've noticed you are using merges, instead of rebase, and the more commit you add on top of merges, the more conflicts you will get when you rebase (and you can also rebase without squashing, to not lose the history).

Anyway... Here are the extra patches:
[patches2.zip](https://github.com/godotengine/godot/files/5612167/patches2.zip)

Some notes:
- 0001: Simply add some known emscripten variables to the linting configuration, so it does not complain.
- 0002: A fixup to my previous patches updating the function signatures :roll_eyes: 
- 0003: The full formatting patch. If this does not apply for some reason, you can run formatting yourself via:
  ```
  cd platform/javascript
  npm ci
  npm run format
  ```
  and then commit differences.

dsnopek commented:
@Faless Thanks! I've added your next set of patches and all checks are passing again.

> I guess that's also up to you. Do you feel it's already usable, and just missing some features? Or do you think it still needs some work to be usable?

It's definitely usable! People are already using it. :-)

That said, I think I would like to finish up adding support for everything (or almost everything) related to VR before taking this out of "Draft" status.

> Rebasing is not urgent, it was more a suggestion for you, since I've noticed you are using merges, instead of rebase, and the more commit you add on top of merges, the more conflicts you will get when you rebase (and you can also rebase without squashing, to not lose the history).

Ah, so, you were just suggesting to use `git rebase` instead of `git merge`, I understand now. :-)

I keep a lot of little branches locally as I work on big changes like this, and if I continually rebased the branch this PR is based on, it would make it harder to merge those branches in later, which is why I've been using merge. Anyway, I think I'll be able to handle the conflicts when I eventually rebase and squash this branch.

dsnopek commented:
Alright!

I've added the last couple things I wanted to add before taking this out of "Draft" status, including adding support for all the events (including select* and squeeze*) such that almost all VR related features are supported, and adding some class reference documentation. There's still a couple small things missing, a couple minor bugs, and, of course, all of AR to implement, but those can be done in later PRs.

So, I've rebased and squashed this whole PR and removed "Draft" status!

@Faless @BastiaanOlij I'd really appreciate your review when you have time! Thanks in advance :-)

dsnopek commented:
PR #44154 is the port to the 'master' branch!

BastiaanOlij reviewed (approved):
I wish I had more time to take this for a proper spin and not just a take a quick glance at the code. 

I love this, absolutely love love love this. Been following dsnopek's progress with his video logs and very impressed by him getting this to work,.

I'm all for merging this in, I'm happy with this being part of the core especially seeing it's all implemented as a module and if people really want to save the few bytes it may spare they can turn the module off.

As webXR is an open standard I don't see anything holding us back here.

dsnopek commented:
@BastiaanOlij Thanks so much!!! :-)

And, here is my next progress report video:

https://www.youtube.com/watch?v=qRW9pNTWcVQ

dsnopek commented:
Per code review from @Faless on IRC, I just added a new commit which completely un-monkey-patches `window.requestAnimationFrame()` when WebXR is uninitialized. I tested it in both Chrome and Firefox with the WebXR emulator extension, and in the Oculus browser on a Quest. All seems to be working!

Faless reviewed (approved):
Well, with the new year I finally had the time to try out the PR with the browser extension.
Once more, this is really impressive work :+1: and I think it's ready to merge.

If you can update the 4.0 PR I'll approve that one too.

dsnopek commented:
@Faless Thanks so much! The 4.0 PR is updated too https://github.com/godotengine/godot/pull/44154

akien-mga commented:
Should the two commits be squashed?

dsnopek commented:
@akien-mga They certainly could be squashed! I didn't squash them in my branch so that it was easy for @Faless to review. You can squash them on merge, or would you like me to squash them?

akien-mga commented:
We don't use squash on merge in the code repo as we want the merged history to match that of PRs (the commit hashes shown here would match what's in `3.2` after merge - if we squash on merge they're dangling commits).

dsnopek commented:
@akien-mga Ah, ok, thanks. In that case, I've squashed the two commits here in the PR (and on the 4.0 one too)!

akien-mga commented:
Thanks and congrats for this great contribution!
